### DISCUSSION TITLE
Connect to GitLab from unleash-proxy via corporate http proxy

### DISCUSSION THREAD

**Discussion by bikerp** [Unleash/unleash#6757] [General]
Hi,
I want to use unleash-proxy to connect to upstream GitLab via corporate proxy. I tried to setup env, variable https_proxy for unleash-proxy but it cannot connect to gitlab. Any ide what could be wrong?

**Comment by sjaanus** [comment]
Hi @bikerp. Can you provide more information on how you set up Unleash? Which environment variable are you setting, and what exactly happens when you try connecting to GitLab? Are you receiving any error messages or codes?

**Comment by bikerp** [comment]
Hi @sjaanus ,
I'm running unleash-proxy as Dcoker conatiner. I provided https_proxy environment variable. The error I see is:
ERROR: FetchError: Unleash Repository error: request to https://<our_gitlab_instance>/api/v4/feature_flags/unleash/91829/client/features failed, reason: getaddrinfo ENOTFOUND <our_gitlab_instance>
I checked the code and I think that I need to provide httpOptions. If I understand it correctly I need to create new nodejs project, add @unleash/proxy dependency and initialize with custom httpOptions

**Reply by sjaanus** [reply]
The error specifically suggests a DNS resolution issue. It seems unable to reach the domain; please check your Docker network configuration.

Another approach would be to run your application outside of Docker, where you know you have a connection to your GitLab. Then, if it fails when placed inside Docker, you'll know it's a Docker network problem.

**Reply by bikerp** [reply]
Yes because I need to use http proxy. Bud the docker image of unleash-proxy doesn't support that out of the box. I solved it as I described above i.e. new nodejs project

**Reply by naderghanbari** [reply]
Running into the same issue, trying to productionize Unleash Proxy in our environment.

@sjaanus I could not find any documentation around "How to configure an http/http proxy when running Unleash Proxy". 

@bikerp  mentioned `https_proxy `, I assume it's a reference to [these node-fetch defaults](https://github.com/node-fetch/node-fetch/issues/195)?

I tried `HTTPS_PROXY` and `HTTP_PROXY` but Unleash Proxy does not seem to respect them. Appreciate any help.

**Reply by chriswk** [reply]
So, according to the issue you linked, node-fetch does not support the proxy environment variables, and one would need to configure something like global-agent before starting the proxy. Though, since unleash-client-node uses [make-fetch-happen](https://github.com/npm/make-fetch-happen?tab=readme-ov-file#opts-proxy) and we don't pass in the opts.proxy variable, I was under the impression from make-fetch-happen's docs that you should be able to use
```
process.env.HTTP_PROXY, process.env.HTTPS_PROXY, and process.env.PROXY
```

**Reply by naderghanbari** [reply]
Thanks for the reply @chriswk .

I tried the env vars you mentioned and Unleash Proxy does not seem to respect them. Ran the app locally and also from Docker with an intercepting proxy, and also once with bad values (expecting toggle fetches to fail), but it made no difference. Requests from Unleash Proxy to the Unleash Server do not seem to go through the proxy. 

I haven't spent much time, but in the `unleash-proxy` -> `unleash-client(node)` -> `make-fetch-happen` -> `@npmcli/agent` dependency chain, there are recent bug fixes that may be pertitnent. https://github.com/npm/agent/pull/67 being the main suspect for me. I will try upgrading to the latest `@npmcli/agent` which fixes that issue and see if that resolved the issue.

If that didn't work, what next steps do you recommend? Opening an issue against the Unleash Proxy or Unleach Client Node projects?

**Reply by naderghanbari** [reply]
Spent some time debugging the code paths locally. Although I am not that familiar with Node, it seems to me that the native support of the `@npmcli/agent` for setting proxies via env vars (implemented in [proxy.js](https://github.com/npm/agent/blob/main/lib/proxy.js)) is masked by the fact that `unleash-client` builds its own agent [here](https://github.com/Unleash/unleash-client-node/blob/main/src/request.ts#L32) so the only way to set the proxy seems to be by passing the [httpOptions](https://github.com/Unleash/unleash-proxy/blob/main/src/config.ts#L43) which does not support env var overrides.

If this assessment is correct, I think `unleash-client-node` should either fix the support for the env vars or document it, and `unleash-proxy` should also document that there is currently no way to set the proxy via environment, and programmatically is the only way.

**Reply by kwasniew** [reply]
This should be fixed in https://github.com/Unleash/unleash-proxy/releases/tag/v1.4.0
